Utils/FaviconFetcher.swift

```diff
@@ -13,108 +16,118 @@ class FaviconFetcherErrorType: ErrorType {
         let deferred = Deferred<Result<[Favicon]>>()
 
         dispatch_async(queue) { _ in
-            if self._favicons.count == 0 {
-                // Initially look for tags in the page
-                self.loadFromDoc()
-            }
-
-            // If that didn't find anything, look for a favicon.ico for this host
-            if self._favicons.count == 0 {
-                self.loadFromHost()
-            }
-
-            var filledCount = 0
-            for (i, icon) in enumerate(self._favicons) {
-                // For each icon we set of an async load of the data (in order to get the width/height.
-                self.getFavicon(icon, profile: profile).upon { result in
+            self.parseHTMLForFavicons(url).bind({ (result: Result<[Favicon]>) -> Deferred<[Result<Favicon>]> in
+                var deferreds = [Deferred<Result<Favicon>>]()
+                if let icons = result.successValue {
+                    deferreds = map(icons) { self.getFavicon(url, icon: $0, profile: profile) }
+                }
+                return all(deferreds)
+            }).bind({ (results: [Result<Favicon>]) -> Deferred<Result<[Favicon]>> in
+                for result in results {
                     if let icon = result.successValue {
-                        self._favicons[i] = icon
+                        oldIcons.append(icon)
                     }
-                    filledCount++
+                }
 
-                    // When they've all completed, we can fill the deferred with the results
-                    if filledCount == self._favicons.count {
-                        self._favicons.sort({ (a, b) -> Bool in
-                            return a.width > b.width
-                        })
+                oldIcons.sort({ (a, b) -> Bool in
+                    return a.width > b.width
+                })
 
-                        deferred.fill(Result(success: self._favicons))
-                    }
-                }
-            }
+                return deferResult(oldIcons)
+            }).upon({ (result: Result<[Favicon]>) in
+                deferred.fill(result)
+                return
+            })
         }
 
         return deferred
     }
 
-    // Loads favicon.ico on the host domain for this url
-    private func loadFromHost() {
-        if let url = NSURL(scheme: siteUrl.scheme!, host: siteUrl.host, path: "/favicon.ico") {
-            let icon = Favicon(url: url.absoluteString!, type: IconType.Guess)
-            _favicons.append(icon)
+    lazy private var alamofire: Alamofire.Manager = {
+        var defaultHeaders = Alamofire.Manager.sharedInstance.session.configuration.HTTPAdditionalHeaders ?? [:]
+        defaultHeaders["User-Agent"] = userAgent
+
+        let configuration = NSURLSessionConfiguration.defaultSessionConfiguration()
+        configuration.timeoutIntervalForRequest = 5
```

Comment:
Infinite AFAICT :) I ran a test here with the Alexa top 1000 to see how many good icons I could find. I set the test timeout at 100s (originally at least) and still wound up timing out.

This is just a random number. Happy for suggestions (but I don't think it should be very long).